File: AndroidTutorial/tutorial/src/main/java/com/pj/tutorial/NativeUIController.kt
package com.pj.tutorial

import android.app.AlertDialog
import com.pj.pubsub.Messenger
import com.pj.pubsub.Tag
import com.pj.pubsub.extensions.ContainerBuilder
import com.pj.pubsub.extensions.Message
import com.pj.pubsub.extensions.add
import com.pj.pubsub.extensions.getString
import com.unity3d.player.UnityPlayer

class NativeUIController{
    private val messenger = Messenger()

    init {
        messenger.setReceivingRule(Tag.native)
        messenger.setBasePublishingTag(Tag.game)
        messenger.subscribe("OPEN_ALERT", ::onReceive)
    }

    private fun onReceive(message : Message){
        // Get data from container
        val alertMessage = message.container.getString("alertMessage")

        openAlert(alertMessage ?: "OPEN_ALERT")
    }

    private fun openAlert(alertMessage: String){
        fun giveBackResult(okPressed : Boolean){
            // Create ContainerBuilder and set data.
            val containerBuilder = ContainerBuilder()
            containerBuilder.add("pressOk", okPressed)
            val reply = Message("ALERT_RESULT", containerBuilder.build())

            // Give back message to message notifier
            messenger.publish(reply)
        }

        val builder = AlertDialog.Builder(UnityPlayer.currentActivity)
        builder.setTitle("Alert")
        builder.setMessage(alertMessage)
        builder.setPositiveButton("OK"){_, _ ->
            giveBackResult(true)
        }
        builder.setOnCancelListener {_->
            giveBackResult(false)
        }

        val dialog = builder.create()
        dialog.show()
    }

}

File: iOSTutorial/iOSTutorial/NativeUIController.swift
//
//  NativeUIController.swift
//  iOSTutorial
//
//  Created by sangmin park on 3/1/24.
//

import Foundation
import iOSPubSub

public class NativeUIController{
    
    private let messageHandler : Messenger
    private let gameViewController : UIViewController
    
    init(gameViewController : UIViewController){
        self.gameViewController = gameViewController
        messageHandler = Messenger()
        messageHandler.setReceivingRule(all: .native)
        messageHandler.setBasePublishingTag(.game)
        messageHandler.subscribe(key: "OPEN_ALERT", handler: onReceive)
    }

    private func onReceive(message: Message){
        let alertMessage = message.container.getString(key: "alertMessage") ?? ""
        openAlert(alertMessage: alertMessage, message: message)
    }

    private func openAlert(alertMessage: String, message: Message){
        func giveBackResult(pressOk: Bool){
            var container = Container()
            container.add(key: "pressOk", value: pressOk)
            let reply = Message(key: "ALERT_RESULT", container: container)
            messageHandler.publish(message: reply)
        }
        
        let alertController = UIAlertController(title:"Alert", message: alertMessage, preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: "OK", style: .default, handler: { _ in
            giveBackResult(pressOk: true)
        }))
        alertController.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: { _ in
            giveBackResult(pressOk: false)
        }))
        print("NativeUIController open alert!")
        gameViewController.present(alertController, animated: true)
        
    }
}
